File: Iteration 2 (XNA)/Source/ProjectileSystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Content;

namespace Asteroids {
	
	class ProjectileSystem {

		private List<Projectile> m_projectiles = null;
		private SoundEffect m_laserSound = null;
		private GameSettings m_settings = null;
		private SpriteSheet m_projectileSprites = null;
		private SpriteAnimation[] m_projectileAnimations = null;

		public ProjectileSystem(SpriteSheetCollection spriteSheets, GameSettings settings, ContentManager content) {
			m_projectiles = new List<Projectile>();

			// load the projectile sprites and create an animation from them
			if(spriteSheets != null) {
				m_projectileSprites = spriteSheets.getSpriteSheet("Projectiles");
			}

			if(m_projectileSprites != null) {
				m_projectileAnimations = new SpriteAnimation[4];
				
				for(int i=0;i<4;i++) {
					m_projectileAnimations[i] = new SpriteAnimation(0.18f, SpriteAnimationType.Loop);
					m_projectileAnimations[i].addSprite(m_projectileSprites.getSprite(i * 2));
					m_projectileAnimations[i].addSprite(m_projectileSprites.getSprite(i * 2 + 1));
				}
			}

			// load the laser sound
			if(content != null) {
				m_laserSound = content.Load<SoundEffect>("Laser");
			}

			m_settings = settings;
		}

		// return the number of projectiles
		public int size() {
			return m_projectiles.Count();
		}

		// return a specific projectile
		public Projectile getProjectile(int index) {
			if(index < 0 || index >= m_projectiles.Count()) { return null; }

			return m_projectiles[index];
		}

		// spawn a projectile at a position with a rotation from a specific space ship
		public void spawnProjectile(Vector2 position, float rotation, SpaceShip spaceShip) {
			if(spaceShip == null) { return; }

			// get the colour of the projectile
			SpriteAnimation projectileAnimation = null;
				 if(spaceShip.colour == SpaceShipColour.Red)    { projectileAnimation = m_projectileAnimations[0]; }
			else if(spaceShip.colour == SpaceShipColour.Blue)   { projectileAnimation = m_projectileAnimations[1]; }
			else if(spaceShip.colour == SpaceShipColour.Green)  { projectileAnimation = m_projectileAnimations[2]; }
			else if(spaceShip.colour == SpaceShipColour.Yellow) { projectileAnimation = m_projectileAnimations[3]; }

			// create the projectile and store it
			m_projectiles.Add(new Projectile(position, rotation, projectileAnimation, spaceShip, m_settings));
			m_laserSound.Play();
		}

		// remove a projectile from the collection
		public bool removeProjectile(Projectile p) {
			if(p == null) { return false; }
			return removeProjectile(m_projectiles.IndexOf(p));
		}

		// remove a projectile at a specific index from the collection
		public bool removeProjectile(int index) {
			if(index < 0 || index >= m_projectiles.Count()) { return false; }
			if(m_projectiles.Count() == 1) {
				m_projectiles.RemoveAt(index);
			}
			else {
				int lastIndex = m_projectiles.Count() - 1;
				Projectile temp = m_projectiles[index];
				m_projectiles[index] = m_projectiles[lastIndex];
				m_projectiles[lastIndex] = temp;
				m_projectiles.RemoveAt(lastIndex);
			}
			return true;
		}

		// update the projectile system
		public void update(GameTime gameTime) {
			// update all of the projectile animations
			for(int i=0;i<4;i++) {
				m_projectileAnimations[i].update(gameTime);
			}
			
			// update all of the projectiles
			for(int i=0;i<m_projectiles.Count();i++) {
				m_projectiles[i].update(gameTime);

				// check for projectiles out of bounds, and remove them
				if(m_projectiles[i].outOfBounds()) {
					int lastIndex = m_projectiles.Count() - 1;
					Projectile temp = m_projectiles[i];
					m_projectiles[i] = m_projectiles[lastIndex];
					m_projectiles[lastIndex] = temp;
					m_projectiles.RemoveAt(lastIndex);
					i--;
				}
			}
		}

		public void draw(SpriteBatch spriteBatch) {
			for(int i=0;i<m_projectiles.Count();i++) {
				m_projectiles[i].draw(spriteBatch);
			}
		}

		public void reset() {
			m_projectiles.Clear();
		}

	}

}

File: Iteration 3 (Supernova Engine)/Source/Game Systems/ProjectileSystem.cpp
#include "Game Objects/SpaceShip.h"
#include "Game Systems/ProjectileSystem.h"

ProjectileSystem::ProjectileSystem()
	: Entity()
	, m_initialized(false)
	, m_laserSound(NULL) {
	for(int i=0;i<SpaceShipColours::NumberOfColours;i++) {
		m_laserAnimations[i] = NULL;
	}
}

ProjectileSystem::ProjectileSystem(const ProjectileSystem & p)
	: Entity(p)
	, m_initialized(p.m_initialized)
	, m_laserSound(p.m_laserSound) {
	for(int i=0;i<SpaceShipColours::NumberOfColours;i++) {
		m_laserAnimations[i] = p.m_laserAnimations[i];
	}

	for(int i=0;i<p.m_projectiles.size();i++) {
		m_projectiles.push_back(new Projectile(*p.m_projectiles[i]));
	}
}

ProjectileSystem & ProjectileSystem::operator = (const ProjectileSystem & p) {
	Entity::operator = (p);

	for(int i=0;i<m_projectiles.size();i++) {
		delete m_projectiles[i];
	}
	m_projectiles.clear();

	m_initialized = p.m_initialized;

	for(int i=0;i<SpaceShipColours::NumberOfColours;i++) {
		m_laserAnimations[i] = p.m_laserAnimations[i];
	}
	
	m_laserSound = p.m_laserSound;

	for(int i=0;i<p.m_projectiles.size();i++) {
		m_projectiles.push_back(new Projectile(*p.m_projectiles[i]));
	}

	return *this;
}

ProjectileSystem::~ProjectileSystem() {
	for(int i=0;i<m_projectiles.size();i++) {
		delete m_projectiles[i];
	}
}

int ProjectileSystem::numberOfProjectiles() const {
	return m_projectiles.size();
}

bool ProjectileSystem::hasProjectile(const Projectile * projectile) const {
	if(projectile == NULL) { return false; }

	for(int i=0;i<m_projectiles.size();i++) {
		if(*m_projectiles[i] == *projectile) {
			return true;
		}
	}
	return false;
}

int ProjectileSystem::indexOfProjectile(const Projectile * projectile) const {
	if(projectile == NULL) { return -1; }

	for(int i=0;i<m_projectiles.size();i++) {
		if(*m_projectiles[i] == *projectile) {
			return i;
		}
	}
	return -1;
}

const Projectile * ProjectileSystem::getProjectile(int index) const {
	if(index < 0 || index >= m_projectiles.size()) { return NULL; }

	return m_projectiles[index];
}

bool ProjectileSystem::addProjectile(Projectile * projectile) {
	if(projectile == NULL) { return false; }

	for(int i=0;i<m_projectiles.size();i++) {
		if(*m_projectiles[i] == *projectile) {
			return false;
		}
	}

	m_projectiles.push_back(projectile);

	return true;
}

bool ProjectileSystem::removeProjectile(int index) {
	if(index < 0 || index >= m_projectiles.size()) { return false; }

	int lastIndex = m_projectiles.size() - 1;
	delete m_projectiles[index];
	m_projectiles[index] = m_projectiles[lastIndex];
	m_projectiles.remove(lastIndex);
	
	return true;
}

bool ProjectileSystem::removeProjectile(const Projectile * projectile) {
	if(projectile == NULL) { return false; }

	for(int i=0;i<m_projectiles.size();i++) {
		if(*m_projectiles[i] == *projectile) {
			int lastIndex = m_projectiles.size() - 1;
			delete m_projectiles[i];
			m_projectiles[i] = m_projectiles[lastIndex];
			m_projectiles.remove(lastIndex);

			return true;
		}
	}
	return false;
}

void ProjectileSystem::clearProjectiles() {
	for(int i=0;i<m_projectiles.size();i++) {
		delete m_projectiles[i];
	}
	m_projectiles.clear();
}

bool ProjectileSystem::init(ResourceManager * resourceManager, SpriteAnimationCollection * animations) {
	if(m_initialized) { return true; }

	if(animations == NULL || animations->numberOfAnimations() == 0) { return false; }

	QString animationName;

	const SpriteAnimation * animation = NULL;

	for(int i=0;i<SpaceShipColours::NumberOfColours;i++) {
		animationName = QString(SpaceShipColours::toString(i)).append(" Laser");

		animation = animations->getAnimation(animationName);
		if(animation == NULL) { return false; }

		m_laserAnimations[i] = animation;
	}

	if(resourceManager != NULL) {
		m_laserSound = resourceManager->getSound("Laser");
	}

	m_initialized = true;

	return true;
}

void ProjectileSystem::spawnLaserBeam(const Vector2 & position, float rotation, const SpaceShip * source) {
	if(!m_initialized || source == NULL) { return; }

	m_projectiles.push_back(new Projectile(m_laserAnimations[static_cast<int>(source->getColour())], source, position, rotation));

	if(m_laserSound != NULL) {
		m_laserSound->play();
	}
}

void ProjectileSystem::reset() {
	if(!m_initialized) { return; }

	clearProjectiles();
}

void ProjectileSystem::update(unsigned int timeElapsed) {
	if(!m_initialized) { return; }

	for(int i=0;i<m_projectiles.size();i++) {
		m_projectiles[i]->update(timeElapsed);
		
		if(m_projectiles[i]->outOfBounds()) {
			int lastIndex = m_projectiles.size() - 1;
			delete m_projectiles[i];
			m_projectiles[i] = m_projectiles[lastIndex];
			m_projectiles.remove(lastIndex);
			i--;
		}
	}
}

void ProjectileSystem::draw() {
	if(!m_initialized) { return; }

	for(int i=0;i<m_projectiles.size();i++) {
		m_projectiles[i]->draw();
	}
}

bool ProjectileSystem::operator == (const ProjectileSystem & p) const {
	return Entity::operator == (dynamic_cast<const Entity &>(p));
}

bool ProjectileSystem::operator != (const ProjectileSystem & p) const {
	return !operator == (p);
}
